src/shm/provider/alloc_layout.rs

```diff
@@ -0,0 +1,175 @@
+//
+// Copyright (c) 2023 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+use std::mem::MaybeUninit;
+
+use crate::{
+    context::{zc_threadsafe_context_t, ThreadsafeContext},
+    errors::z_error_t,
+    shm::protocol_implementations::posix::posix_shared_memory_provider::PosixAllocLayout,
+    transmute::{
+        unwrap_ref_unchecked, Inplace, TransmuteIntoHandle, TransmuteRef, TransmuteUninitPtr,
+    },
+    z_owned_buf_alloc_result_t,
+};
+use libc::c_void;
+use zenoh::shm::{
+    AllocLayout, BlockOn, Deallocate, Defragment, DynamicProtocolID, GarbageCollect, JustAlloc,
+};
+
+use crate::context::Context;
+
+use super::{
+    alloc_layout_impl::{alloc, alloc_async, alloc_layout_new},
+    shared_memory_provider::z_loaned_shared_memory_provider_t,
+    shared_memory_provider_backend::DynamicSharedMemoryProviderBackend,
+    types::z_alloc_alignment_t,
+};
+
+/// A loaned SharedMemoryProvider's AllocLayout
+#[cfg(target_arch = "x86_64")]
```

Comment:
Can these be generated as opaque types ?

Reply:
No, because it uses types defined in zenoh_c library (we cannot use zenoh_c types in opaque_types because zenoh_c technically depends on opaque_types)

Reply:
Can we  define them directly in terms of zenoh_c types? Alternatively, we can probably define some placeholders corresponding to needed zenoh_c_types and generate opaque types from them? I mean, can we get rid of explicit type size dependency on target arch ?

Reply:
I will think about this.